streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignorTest.java

```diff
@@ -655,6 +638,41 @@ public void shouldAssignTasksToNewClientWithoutFlippingAssignmentBetweenExisting
         assertThat(newClient.activeTaskCount(), equalTo(2));
     }
 
+    @Test
+    public void shouldAssignTasksToClientsWithImprovedBalancing() {
+        final ClientState c1 = createClient(p1, 6);
+        final ClientState c2 = createClient(p2, 6);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02, task03, task04, task05);
+        taskAssignor.assign(1);
+
+        assertTrue(Math.abs(c1.getNumberOfActivePartitions() - c2.getNumberOfActivePartitions()) <= 4);
+        assertTrue(Math.abs(c1.getNumberOfActiveStateStores() - c2.getNumberOfActiveStateStores()) <= 4);
+        assertTrue(Math.abs(c1.getNumberOfStandbyStateStores() - c2.getNumberOfStandbyStateStores()) <= 4);
+    }
+
+    @Test
+    public void shouldAssignMultipleReplicasToClientsWithImprovedBalancing() {
+        createClientWithPreviousActiveTasks(p1, 1, task00);
+        createClientWithPreviousActiveTasks(p2, 1, task01);
+        createClientWithPreviousActiveTasks(p3, 1, task02);
+        createClientWithPreviousActiveTasks(p4, 1, task03);
+        createClientWithPreviousActiveTasks(p5, 1, task04);
+        createClientWithPreviousActiveTasks(p6, 1, task05);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02, task03, task04, task05);
+        taskAssignor.assign(2);
+
+        //do not expect too accurate behavior, there is still some limits to our approximation of the "capacity" of a task
+        for (final int processId : clients.keySet()) {
+            if (processId < p1 || processId > p6) {
+                continue;
+            }
+            assertTrue(clients.get(processId).getNumberOfStandbyStateStores() < 9 &&
```

Comment:
I could be missing something but a range of 1 to 9 state stores seems large to me for an individual task to have

Reply:
Yeah, I thought so to. Although with our current design, it appears that this is what we end up with. When I printed the number of state stores for each ClientState, it ran from 2 to 8 (inclusive). We might need to change the assignment strategy somehow.